### ISSUE TITLE
Fast Crash Recovery times

### ISSUE THREAD

**Issue by dd6905** [ReturnInfinity/BMFS#57]
I've implemented a proprietary solution to utilize sofware to recovered a crashed file system from matter of a week to a day. can have extensions to utilize utilize additional DMA hw on server, and/or RDMA support. for fast crash recovery.

**Comment by tay10r** [comment]
This is really ambiguous

**Comment by dd6905** [comment]
There are several parts here, but I just wanted to lump it into one placeholder until we have a chance to talk more later and I get more of a sense of the roadmap and priorities and what is within scope. Maybe a better title might have been "Direct or Fast Creation/Insertion Path with different priority in performance for bulk transfers" or "Bulk Import via fast path bypassing regular file system file creation for speedier recovery" and "extensible HAL for hardware assist in bulk trasfers w/o CPU interaction (we are talking about addressing both issues created from within same server scenario, and between server scearios)". When recoverying file by file in a large system recovery sequentially we found that the file system file creation was our biggest bottleneck  and not being able to initially parallize the file creation process without some creative file management and required more man hour overhead. If we could recover all metadata first, then data later asynchronously prioritized on files that were are dirty ad actually requested/needed by users/engineers (without a transfer of knowedge from user to IT). My job was to reduce recovery from week+ to wihin 72, 24, 8, 4, 2, 1 hour, criteria so most urgent was recovered within one business day. Only the IT lost sleep. and we did come up with ways to allow engineers to start retrieving files for offline editing and automatic merging when the file system was fully up. This had different challenges as the data got larger and larger having exceeded a petabyte onward to exabyte support.  I do not yet understand enough of this projects design history to say much more than vague at this time, but having worked on this for many years and having a kludged together solution utilizing specific equipment, I'd like to eventually see a clean redesign of hooks (independent of specific custom hardware using an hardware adaptation layer) for fast bulk file metadata insertion (I even propose two independent paths - one for metadata to present a coherent file system first to users earlier( giving the impression that the file system was up and running) (and being able to even manipulate file system metadata eg renames, moves, etc without the back end file content actually existing) and usable and another fast path for file data insertion (which can actually be done asynchronously, prioritized on any file that is demanded by a user) with hardware assist feature to be able to design a fast insertion path being able to utilizing any additonal hardware to assist in faster file insertion path that does not require CPU interaction. We decided to utilize copy-on-write so we could even use parallize recovery from two different sets of different backups since most of the data between backup sets were identical. My conclusion, which I'm working on right now, was to utilize a best of both worlds design gaining speed of fixed length transfers yet support variable lengh blocks without the penalty (eg memory management issue kludges) similarily like pipelining. All transfers were cell or fixed length speeding up network transfers and being able to parallize transfers, but still allow for support for heterogenus environment of different block sizes. Hence the higher priority need for a variable per file block size feature. I currently require TCP transfers so the network transport layer handles out of order packet transfers. Switching to support UDP would increase performance, and the re-ordering issue can be handled at a lower priority as well as lost packets -- just mark them as dirty, and file system can lower priority on having to re-retrieve data until file system marks that data as being needed. During some kind of major failure eg geographic site wide failure, and having to restore 100% data from backups. We could truck drives back, but if any corruption existed in the backups, we had to always do a clean recovery through the file system. We found that the file system was the biggest bottleneck in recovery time. We found ways to bypass some of it, but it ended up being custom to our specialized hardware. I  wanted to keep a placeholder to eventually add features in to the file system to be optimized at data recovery at near line speeds also hooks to be able to utilize any various additional hardware that might be plugged in to the server via RAID hardware and 10/40/100GigabitEtherne with RDMA support. Some form of hooks and HAL/hardware adaptation layer in the design to bypass seamlessly utilize the DMA or RDMA features in the fastpath. More details better saved for a different  design forum. It is also true that some of this can be in the file system metadata format, some in the file system api, and all in third party exernal tools, but placing support for these features at lower levels ultimately reduces the total code required.
